ADBDEV-7618: Change General to OuterQuery locus in bring_to_outer_query()

mos65o2 opened the pull request:
Change General to OuterQuery locus in bring_to_outer_query (#1691)

We assume that if we have a correlated subquery for a given path we process
the OuterQuery locus. The patch 605f84e broke this logic - this left the
General locus, even for the correlated subplan (main problem). So we revert it.
Another issue was that the Values ​​Scan node was not treated as OuterQuery
with parameterized qual. This resulted in invalid plans with parameterized
qual in Values ​​Scan if Motion was added on top.
The solution is to allow restrictinfo to be moved to upperrestrictinfo
(see definition) instead of baserestrictinfo for parameterized Values Scan.
Since upperrestrictinfo is not empty, this calls the bring_to_outer_query
handler to distribute the data and prevents to add Motion above the
parameterized node.

mos65o2 committed:
Revert "Do not bring to outerquery locus for GeneralLocus path."

This reverts commit 605f84e62144348673463e61799f5c7edd16ec11.

Stolb27 reviewed (commented):
We discussed the exact case to test this code block several days ago. Why did you not append it?

mos65o2 commented:
> We discussed the exact case to test this code block several days ago. Why did you not append it?

It looks like this case is not testing `bring_to_outer_query` with the General locus. It actually has the Hashed locus.
And it seems that it is impossible to reproduce such a parameterized subplan so that Motion is added to the top without a volatile function.
Discussed case:
```SQL
set optimizer = 0;
create table t1 (i int, j int) distributed by (i);
insert into t1 select i, i from generate_series(1, 10) i;
analyze t1;

create table t_hashed (i int, j int) distributed by (i);
insert into t_hashed values (1,1), (2,2), (3,3);
analyze t_hashed;

explain (verbose, costs off) select * from t1 where t1.j > (select i from t_hashed, ( values (1), (2), (3)) k (a)  where a < t1.i );
                              QUERY PLAN                               
-----------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)
   Output: t1.i, t1.j
   ->  Seq Scan on public.t1
         Output: t1.i, t1.j
         Filter: (t1.j > (SubPlan 1))
         SubPlan 1
           ->  Materialize
                 Output: t_hashed.i
                 ->  Broadcast Motion 3:3  (slice2; segments: 3)
                       Output: t_hashed.i
                       ->  Nested Loop
                             Output: t_hashed.i
                             ->  Seq Scan on public.t_hashed
                                   Output: t_hashed.i, t_hashed.j
                             ->  Values Scan on "*VALUES*"
                                   Output: "*VALUES*".column1
                                   Filter: ("*VALUES*".column1 < t1.i)

```

mos65o2 committed:
Change the need to separate restrictinfo for RTE_VALUES

If the Values ​​Scan of a subquery has parameters, the locus of that path
must change from General to OuterQuery. The locus change occurs if
upperrestrictinfo is not empty (see set_rel_pathlist()). In order to
fill upperrestrictinfo for Values ​​Scan when it has parameters,
rel_need_to_separate_outer_query_restrictinfos() must return "true".
Before this commit, this did not happen and plans with parameterized
Values ​​Scan could fail because Motion could be added above that node.

Stolb27 commented on src/backend/optimizer/path/allpaths.c:
```diff
@@ -522,10 +522,7 @@ bring_to_outer_query(PlannerInfo *root, RelOptInfo *rel, List *outer_quals)
 		Path	   *path;
 		CdbPathLocus outerquery_locus;
 
-		if (CdbPathLocus_IsGeneral(origpath->locus) ||
```
Do we really want to apply the common flow of transformation from General to Outer locus here? Motion isn't needed, as Material and as Result as consequence. Is the useless Result node in plans below caused by this decision? Maybe do we simply replace locus for this case instead of leaving it as is, as was done before.

mos65o2 replied:
> Do we really want to apply the common flow of transformation from General to Outer locus here?

The current implementation of `cdbpath_create_motion_path` does not add Motion. It just [changes](https://github.com/arenadata/gpdb/blob/513b8d1f123eacd1742172377f24efda6abc8c0e/src/backend/cdb/cdbpath.c#L230) the subpath locus. So the plan remains the same.

> Is the useless Result node in plans below caused by this decision?

Partly yes. Because now outer_quals (upperrestrictinfo) [is not empty](https://github.com/arenadata/gpdb/blob/513b8d1f123eacd1742172377f24efda6abc8c0e/src/backend/optimizer/path/allpaths.c#L771).

Maybe it really is worth just changing the locus and adding a path to the relation without projection. This does not break the tests and output is:
```
explain (costs off) select * from table1 where table1.j >
    (select i from table2, (values (1), (2), (3)) k (a) where a > table1.i limit 1);
                                QUERY PLAN                                
--------------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)
   ->  Seq Scan on table1
         Filter: (j > (SubPlan 1))
         SubPlan 1
           ->  Limit
                 ->  Nested Loop
                       ->  Values Scan on "*VALUES*"
                       ->  Materialize
                             ->  Gather Motion 3:1  (slice2; segments: 3)
                                   ->  Seq Scan on table2
 Optimizer: Postgres-based planner
```

mos65o2 replied:
It looks like we don't need to add the Result node [here](https://github.com/arenadata/gpdb/blob/513b8d1f123eacd1742172377f24efda6abc8c0e/src/backend/optimizer/plan/createplan.c#L2060-L2106) if we have the same TargetLists (needs_result_node == false). We only need to assign qual to the subplan.

mos65o2 replied:
We need a different criterion to evaluate adding a Result node because it breaks the tests.

mos65o2 replied:
An additional criterion for Result exclusion and pushdown filter can be a check that the underlying node is a Values ​​Scan.

On the other hand, this looks like a common approach to filtering with OuterLocus - the filter should be in Result. Scan -> Motion? -> Materilize? -> Result. For example, this plan is generated by the same transformation flow. The filter with the parameter in Result seems appropriate, since we could not exclude Motion:
```
explain ( verbose, costs off) select * from t1 where t1.j > (select t_hashed.i from t_hashed, val  where val.i < t1.i);
                             QUERY PLAN                             
--------------------------------------------------------------------
 Gather Motion 1:1  (slice1; segments: 1)
   Output: t1.i, t1.j
   ->  Seq Scan on public.t1
         Output: t1.i, t1.j
         Filter: (t1.j > (SubPlan 1))
         SubPlan 1
           ->  Nested Loop
                 Output: t_hashed.i
                 ->  Result
                       Output: val.i
                       Filter: (val.i < t1.i)
                       ->  Materialize
                             Output: val.i
                             ->  Seq Scan on public.val
                                   Output: val.i
                 ->  Materialize
                       Output: t_hashed.i
                       ->  Gather Motion 3:1  (slice2; segments: 3)
                             Output: t_hashed.i
                             ->  Seq Scan on public.t_hashed
                                   Output: t_hashed.i
 Optimizer: Postgres-based planner
```


On the other hand, filtering in the scan node or in the upstream Result node by cost is not so critical, since the cost will be approximately the same (cpu_tuple_cost * rows + evaluating the tlist).

Stolb27 replied:
> The filter with the parameter in Result seems appropriate, since we could not exclude Motion

Did you mean Material? The Motion was excluded in this case and Material with Result are useless too.

Ok. If there is no short path to push back quals to the baserestrictinfo let's leave it for further optimizations.

mos65o2 committed:
Merge branch 'adb-7.2.0' into ADBDEV-7618

mos65o2 committed:
Edit test

Stolb27 reviewed (commented):
Please provide more high-level description. You should not describe code lines. You should tell about meaning of patch.

mos65o2 committed:
Merge branch 'adb-7.2.0' into ADBDEV-7618
